### PR TITLE
OHCS-84 add request exam function in the laboratory section

### PR DESCRIPTION
**Pull request by SilverD3**
Changes and additions:

## Laboratory exams filter
- Add column `status` in laboratory exams table;
- Add filter `status` in laboratory exams;
- Adjust date filter: disable future date, set `dateFrom` and `dateTo` according to selected month and year;
- Adjust date filter: Fix hours, minutes and seconds of `dateFrom` to 0 and for `dateTo`, fix hours to 23 and minutes to 59;

## Lab exam edition
- Require material;
- Set `DONE` status when exam results are provided; 
- Remove Lab exam form in patient activities

## Add exam request
- Add exam request form in laboratory section of patient dashboard;
- Add requested exams in laboratory section of patient dashboard;
- Add exam status change;
- Cancel exam request

### CODE REVIEW

**Comment by axelpezzo** [comment]
@SilverD3 I would ask you those things:
- What is exactly the purpose of the "request an exam" inside the laboratory form?
- I saw that the API's are not rady yet, does the backend team already know about this?

**Comment by SilverD3** [comment]
Hello @axelpezzo ! For your questions :

- The purpose of the "request an exam" in the laboratory exam form is to "provide a way to a doctor to request an exam for a patient while doing visit", that's what Alberto Mandelli said. And you can see the description of the Jira ticket [here (OHCS-84)](https://openhospital.atlassian.net/jira/servicedesk/projects/OHCS/queues/custom/25/OHCS-84) for more explanation please.
- Yes, the feature is already present in the backend on the develop branch, but not yet deployed in the test environment as the backend team is facing some issue with Dockerfile.

File: src/components/accessories/laboratory/Exams.tsx [hunk 1]
```diff
@@ -85,16 +110,35 @@ export const Exams: FC = () => {
   const errorMessage = useSelector((state: IState) =>
     state.laboratories.searchLabs.error?.message
       ? state.laboratories.searchLabs.error?.message
+      : t("common.somethingwrong") ||
+        state.laboratories.updateLab.error?.message
+      ? state.laboratories.updateLab.error?.message
```
Comment (axelpezzo) [inline src/components/accessories/laboratory/Exams.tsx:115]: Please use a different variable in order to mantein the code as clean as possibile.

Plus, add the message inside the mock when the api fails (now seems to be empty).

File: src/components/accessories/laboratory/examForm/ExamForm.tsx [hunk 2]
```diff
@@ -349,6 +417,31 @@ const ExamForm: FC<ExamProps> = ({
...
+              <AutocompleteField
+                fieldName="status"
+                fieldValue={formik.values.status}
+                label={t("lab.status")}
+                isValid={isValid("status")}
+                errorText={getErrorText("status")}
+                onBlur={onBlurCallback("status")}
+                options={statusOptions}
+                disabled={isLoading}
+              />
```
Comment (axelpezzo) [inline src/components/accessories/laboratory/examForm/ExamForm.tsx:443]: We should turn this components into a simple select list.

File: src/components/accessories/laboratory/filter/ExamFilterForm.tsx [hunk 3]
```diff
@@ -265,6 +320,17 @@ export const ExamFilterForm: FC<IExamFilterProps> = ({ fields, onSubmit }) => {
...
+                <div className="filterLabForm__item col-3">
+                  <AutocompleteField
+                    fieldName="status"
+                    fieldValue={formik.values.status}
+                    label={t("lab.status")}
+                    isValid={isValid("status")}
+                    errorText={getErrorText("status")}
+                    onBlur={onBlurCallback("status")}
+                    options={examStatusOptions}
+                  />
```
Comment (axelpezzo) [inline src/components/accessories/laboratory/filter/ExamFilterForm.tsx:332]: Selectlist here too.

File: src/components/accessories/patientExams/ExamForm/ExamForm.tsx [hunk 4]
```diff
@@ -242,6 +296,31 @@ const ExamForm: FC<ExamProps> = ({
...
+              <AutocompleteField
+                fieldName="status"
+                fieldValue={formik.values.status}
+                label={t("lab.status")}
+                isValid={isValid("status")}
+                errorText={getErrorText("status")}
+                onBlur={onBlurCallback("status")}
+                options={statusOptions}
+                disabled={isLoading}
+              />
```
Comment (axelpezzo) [inline src/components/accessories/patientExams/ExamForm/ExamForm.tsx:322]: Selectlist here too.

**Comment by axelpezzo** [comment]
@SilverD3 there is still something that not convince me yet about the status of this PR.

I mean, for what I got from the discussion about you and Domanico, I understand that the status of the Exams should be changed behind a controlled actions:

- Doctor makes a request → exam request in DRAFT
- Laboratorist takes in charge the request (new button) → OPEN
- Laboratorist inserts the result (pencil button) → DONE
- Laboratorist deletes (trash bin) either a DRAFT, OPEN or DONE → INVALID or physical deletion (to decide)

If this is corrects I would change the code:
- remove the status field inside the Exam Form
- add the status change logic based on the discussed workflow.

Plus, the "request exam" list actually is pretty useless in due to the fact that the user can't interact with that once a new exams request is created. For what I undestand, when a user create a new "request exam" a new entry should be inserted with no data inside the results AND a draft status.

Inside the "request exam" list I would add a new action button that turns the exams status in "OPEN" and the transfer the row from the "request exam" list to the "exam" list with the updated status.

Then, when a user updates the "exam" inserting the results, the status should be updated too in "DONE".

File: src/components/accessories/laboratory/examForm/ExamForm.tsx (hunk 2 above)
Comment (SilverD3) [reply src/components/accessories/laboratory/examForm/ExamForm.tsx:443]: I didn't edit this field. Also, there could be more twenty exams, so an autocomplete field is suitable.

File: src/components/accessories/laboratory/filter/ExamFilterForm.tsx (hunk 3 above)
Comment (SilverD3) [reply src/components/accessories/laboratory/filter/ExamFilterForm.tsx:332]: Okay

File: src/components/accessories/patientExams/ExamForm/ExamForm.tsx (hunk 4 above)
Comment (SilverD3) [reply src/components/accessories/patientExams/ExamForm/ExamForm.tsx:322]: Removed

**Comment by mwithi** [comment]
FWIW, I found the @axelpezzo understanding matching the requirements and his explanation very clear to me.

I confirm, like @SilverD3 says, that backend changes are ready on develop branches as per OH2-183 (done)

Anyway Dockerfile issues should have been resolved with #469 and #470, am I right @alefalezza ?

**Comment by axelpezzo** [comment]
Hello @SilverD3 few notes:

Logical changes: 

- When a "requested exam" is DRAFT we should show the cancel and the "open" button in order to convert it into a OPEN exam. When this button is clicked we should display a confirmationModal previously.
- When a "requested exam" is OPEN we should display the cancel / edit button. The edit button leads me to the edit form. 
- When in the edit form we insert a result and we click on "save", we should display a confirmationModal in order to warning the user that the exams will be put into a DONE status.
- When a "requested exam" is DONE we should display the edit / delete (logically) button.
- When a "requested exam" is CANCELLED we should display no buttons.

UI changes:

- Align the request an exam form with the OH UI (e.g the button shouldn't be outlined without the icon and when the form is submitted we should display a confirmationModal and not an inline message).

**Comment by axelpezzo** [comment]
@SilverD3 did you check my comment below? https://github.com/informatici/openhospital-ui/pull/462#issuecomment-1571561474

**Comment by SilverD3** [comment]
For patient activities dashboard (lab section), I don't think we should display "open" button, given that a doctor can't take in charge an exam and that section if for doctors (not laboratists, if I'm right)

**Comment by axelpezzo** [comment]
I agree with both of you. @SilverD3 please tell me when it's done for you in order to revew the PR and merge it. I think we're close.

**Comment by SilverD3** [comment]
> For patient activities dashboard (lab section), I don't think we should display "open" button, given that a doctor can't take in charge an exam and that section if for doctors (not laboratists, if I'm right)

Please @axelpezzo , I'm waiting your answer.

**Comment by mwithi** [comment]
> > For patient activities dashboard (lab section), I don't think we should display "open" button, given that a doctor can't take in charge an exam and that section if for doctors (not laboratists, if I'm right)
> 
> Please @axelpezzo , I'm waiting for your answer.

@SilverD3, Alex already agreed with you ("I agree with both of you")

**Comment by SilverD3** [comment]
@mwithi @axelpezzo PR ready to be merge

File: api/oh.yaml [hunk 5]
```diff
@@ -8,8 +8,8 @@ info:
   license:
     name: Apache 2.0
     url: https://www.apache.org/licenses/LICENSE-2.0
-host: "oh2.open-hospital.org"
-basePath: /oh-api
+host: "localhost:8080"
```
Comment (mwithi) [inline api/oh.yaml:11]: is this change meant to be? for me would very good but... you find the way to customize this parameter for different env (develop, test, production, etc...)?

File: api/oh.yaml [hunk 6]
```diff
@@ -94,7 +94,7 @@ tags:
   - name: ward-controller
     description: Ward Controller
 schemes:
-  - https
+  - http
```
Comment (mwithi) [inline api/oh.yaml:97]: is this change meant to be? for me would very good but... you find the way to customize this parameter for different env (develop, test, production, etc...)?

**Comment by mwithi** [comment]
Please @SilverD3 could you fix the conflicts on this PR?

**Comment by mwithi** [comment]
> Please @SilverD3 could you fix the conflicts on this PR?

Up! 📤

**Comment by SilverD3** [comment]
> > Please @SilverD3 could you fix the conflicts on this PR?
> 
> Up! 📤

Hello @mwithi ! Is `develop` branch of the backend up to date ? So I can continue resolving conflicts on this PR.

**Comment by mwithi** [comment]
> > > Please @SilverD3 could you fix the conflicts on this PR?
> > 
> > 
> > Up! 📤
> 
> Hello @mwithi ! Is `develop` branch of the backend up to date ? So I can continue resolving conflicts on this PR.

yes it is update

**Comment by SilverD3** [comment]
@mwithi @axelpezzo conflicts fixed !

**Comment by mwithi** [comment]
I've inserted a draft request but it is not shown in Laboratory section.

Also, the date filter is keeping setting itself automatically to tomorrow date.

image

**Comment by SilverD3** [comment]
@mwithi 

> I've inserted a draft request but it is not shown in Laboratory section.

There is an issue with backend pagination, I think. Also you're not seeing the exam request you created because backend doesn't sort data by most recent. I reported the bug to Arnaud.

> Also, the date filter is keeping setting itself automatically to tomorrow date.

Parsing date with timezone is a little complicated. Anyway, I found a way to fix it.

**Pull request merged**